klippa/src/hmtx.rs

```diff
@@ -0,0 +1,79 @@
+//! impl subset() for hmtx
+
+use crate::{Plan, Subset, SubsetError};
+use read_fonts::TopLevelTable;
+use std::cmp;
+use write_fonts::tables::{hmtx::Hmtx, hmtx::LongMetric};
+
+impl Subset for Hmtx {
+    fn subset(&mut self, plan: &mut Plan) -> Result<bool, SubsetError> {
+        let gids = &plan.glyph_ids;
+        if gids.is_empty() {
+            Err(SubsetError::SubsetTableError(Hmtx::TAG))
+        } else {
+            let num_glyphs = self.h_metrics.len() + self.left_side_bearings.len();
+            let mut num_long_metrics = cmp::min(num_glyphs, 0xFFFF_usize);
+            let last_advance = get_gid_advance(&self.h_metrics, num_glyphs - 1);
+            for gid in gids.iter().rev() {
+                let glyph_id = gid.to_u16() as usize;
+                // skip invalid gids
+                if glyph_id >= num_glyphs {
+                    continue;
+                }
+
+                let advance = get_gid_advance(&self.h_metrics, glyph_id);
+                if num_long_metrics == 1 || last_advance != advance {
+                    break;
+                }
+                num_long_metrics -= 1;
+            }
+
+            let mut new_metrics = Vec::<LongMetric>::with_capacity(num_long_metrics);
+            let mut new_side_bearings = Vec::<i16>::new();
+            for gid in gids {
+                let glyph_id = gid.to_u16() as usize;
+                // skip invalid gids
+                if glyph_id >= num_glyphs {
+                    break;
+                }
```

Comment:
what do we know about the `gids` array? we're doing a lot of checking for invalid values, but would it be possible to validate ahead of time that all the gids in the plan exist in the font?

Reply:
check and filter out invalid gids and then do old gid->new gid mapping, all these should happen during planning phase. So this is just temporary code for setting up the crate and will be revisited later.